Detect coincident points when drawing arcs with rounding.

jasondavies opened the pull request:
When rounding is used, it's possible for `arc()` to generate empty arcs in the case where the start and end points are almost coincident, and become coincident after rounding is applied.

This adds a check for coincident points after rounding is applied, and splits the arc into two if coincident points are detected.

Fixes #38.

jasondavies committed:
Detect coincident points when drawing arcs with rounding.

When rounding is used, it's possible for `arc()` to generate empty arcs
in the case where the start and end points are almost coincident, and
become coincident after rounding is applied.

This adds a check for coincident points after rounding is applied, and
splits the arc into two if coincident points are detected.

Fixes #38.

Fil commented on src/path.js:
```diff
@@ -23,12 +23,33 @@ function appendRound(digits) {
   };
 }
 
+function round(digits) {
+  const k = 10 ** digits;
+  return function(value) {
+    return Math.round(value * k) / k;
+  };
+}
+
+function equal(x0, y0, x1, y1) {
+  return x0 === x1 && y0 === y1;
+}
+
+function equalRound(digits) {
+  let d = Math.floor(digits);
+  if (d > 15) return equal;
+  const r = round(digits);
+  return function(x0, y0, x1, y1) {
+    return r(x0) === r(x1) && r(y0) === r(y1);
+  };
+}
+
```
```suggestion
function equal(x0, y0, x1, y1) {
  return x0 === x1 && y0 === y1;
}

function equalRound(digits) {
  const d = Math.floor(digits);
  if (d > 15) return equal;
  const k = 10 ** d;
  const r = function(value) {
    return Math.round(value * k);
  };
  return function(x0, y0, x1, y1) {
    return r(x0) === r(x1) && r(y0) === r(y1);
  };
}
```

Skip the division, that's 5% ~ 10% faster. Use `d` rather than `digits` to compute the power of 10.

Note: the "round" function does not round anymore, but we inline it so we don't have to think of a different name.

jasondavies committed:
Add changes suggested by @Fil.

Skip unnecessary division for performance. Use `d` rather than `digits` to compute the power of 10.

Co-authored-by: Philippe Rivière <fil@rezo.net>

Fil commented on src/path.js:
```diff
@@ -23,12 +23,28 @@ function appendRound(digits) {
+  return function(x0, y0, x1, y1) {
+    return r(x0) === r(x1) && r(y0) === r(y1);
+  };
+}
```
```suggestion
}

```

(my mistake)

Fil committed:
restore newline

mbostock commented on src/path.js:
```diff
@@ -23,12 +23,29 @@ function appendRound(digits) {
     this._x1 = this._y1 = null; // end of current subpath
     this._ = "";
     this._append = digits == null ? append : appendRound(digits);
+    this._equal = digits === null ? equal : equalRound(digits);
```
```suggestion
    this._equal = digits == null ? equal : equalRound(digits);
```

Use loose equals to match previous line.

mbostock commented on src/path.js:
```diff
@@ -23,12 +23,29 @@ function appendRound(digits) {
+
+function equalRound(digits) {
+  const d = Math.floor(digits);
+  if (d > 15) return equal;
```
```suggestion
  if (!(d <= 15)) return equal;
```

Use pessimistic check in case of NaN.

mbostock reviewed (changes requested):
Thanks for the fix! And hi Jason! 👋 Two small nits.

jasondavies committed:
Add suggestions from @mbostock:

- Use pessimistic `!(d <= 15)` check in case of NaN.
- Use `digits == null` check to match previous line.

jasondavies commented:
It might also be worth generating a set of visual tests to double-check ccw={true, false} and values of a0, a1 where abs(a1 - a0) is some multiple of 2π, in case there's a faulty assumption with da /= 2.

jasondavies commented:
I think my assumptions are correct. After these lines:

```js
// Does the angle go the wrong way? Flip the direction.
if (da < 0) da = da % tau + tau;

// Is this a complete circle? Draw two arcs to complete the circle.
if (da > tauEpsilon) {
  this._append`A${r},${r},0,1,${cw},${x - dx},${y - dy}A${r},${r},0,1,${cw},${this._x1 = x0},${this._y1 = y0}`;
}
```

We can be sure that 0 ≤ da ≤ tauEpsilon, hence it's safe to use da /= 2 to determine the midpoint of the arc.

jasondavies commented:
Ping :)

curran commented:
Great to hear from you @jasondavies !!! I must say I'm a huge fan of your work. This contribution looks amazing.

jasondavies commented:
Ping @mbostock @Fil 👋
